### PR TITLE
Modals: Quantity (complete), rewards (initial design)

### PR DESCRIPTION
**Pull request by anniero98**
# What's new in this PR
Quantity modal replacing the `addToCart` and `removeFromCart` basic functionalities. 
Can be triggered by either `ProductDisplayCard` or `LineItemCard`; but the modal for a single product showing is mutually exclusive between the two. If none of that product has been added to the cart, you can click the product display card. Otherwise, you should tap the line item card to adjust quantity. This might change later, but is a very small change in # of lines of code.

Rewards modal from the initial design is also fully implemented in this PR (originally a dependent PR).

[//]: # "Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

### Online sources
@aivantg was my online source for how to restructure the duplicated modals that show/hide on a condition + wrap different modal-trigger components. 

[//]: # "Optional - copy links to any tutorial or documentation that was useful to you when working on this PR"

### Related PRs

[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

#13 (replacing - closed with comments). Note: rewards modal design has updated. Because the checkout flow is fully functional with the code in this PR, I think we can go ahead and merge the two modals and I'll update the rewards modal according to the new designs in another PR.

## How to review

Styling: 
(Checkout)
I fixed some bugs with the alignment / spacings in Checkout by using width/height percentages instead of raw pixel values. This is again due to the need for responsiveness. It works fine on the simulator, but we should test to make it more robust. I also don't know if someone else has made their own fix for this in their current branch.

(Modal)
- (hacky) Having the icon stick in the top-left corner was tricky for me - ended up using `align-self` and wrapping the rest in a container  with `flex: 1` to have it grow. Don't know how it works exactly still but was pretty convinced by the internet that this is where you should actually use `flex: 1`. Could be wrong anyway. See "Media objects" https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Flexible_Box_Layout/Typical_Use_Cases_of_Flexbox
- (hacky) I used `space-around` for the icon container and the rest of content, but there always looked like there was an unnaturally large gap between them. I couldn't figure out why, but using `top: -20` while giving the icon `padding: 20` on all sides but `paddingBottom` made it look pretty normal. 
image
image
^ Here's what the spacing looks like right now
- Figma spacing also seemed unlikely to work because modals are centered responsively (i.e rely on width and height percentages), so many of the numbers are what I got from trial and error, loosely based on the Figma px values.. 

[//]: # "The order in which to review files and what to expect when testing locally"

## Next steps
Re-implement rewards modal

[//]: # "What's NOT in this PR, doesn't work yet, and/or still needs to be done"

## Tests Performed, Edge Cases

[//]: # "Hopefully we will add a testing suite/CI soon, but until then note down the steps you took to test locally"

### Screenshots
image
image
You will notice that it is left-aligned for the `Title`! because I could not tell from the Figma, and for the longer titles it looks left-aligned ... but it's a one-css-(selector)? fix

[//]: # "Add screenshots of expected behavior - GIFs if you're feeling fancy!"

CC: @wangannie @thumn 

[//]: # "This tags in both Annies as a default. Feel free to change, or add on anyone who you should be in on the conversation."

### CODE REVIEW

File: components/LineItemCard.js [hunk 1]
```diff
@@ -1,33 +1,29 @@
 import React from 'react';
-import { View } from 'react-native';
-import { LineItem, LineItemRow } from '../styled/checkout.js';
 import Colors from '../assets/Colors';
-import { Subhead, Body } from '../components/BaseComponents';
 import { calculateLineItemPrice } from '../lib/checkoutUtils';
+import { LineItem, LineItemRow } from '../styled/checkout';
```
Comment (anniero98) [inline components/LineItemCard.js]: VS Code settings: organizeImports ;_;

File: components/ProductDisplayCard.js [hunk 2]
```diff
@@ -1,8 +1,8 @@
 import React from 'react';
 import { Image } from 'react-native';
-import { ProductCard } from '../styled/checkout.js';
-import { Body, Caption } from '../components/BaseComponents';
 import Colors from '../assets/Colors';
+import { ProductCard } from '../styled/checkout';
```
Comment (anniero98) [inline components/ProductDisplayCard.js]: organize imports pt 2 I've left a message in Slack

File: components/BaseComponents.js [hunk 3]
```diff
@@ -52,7 +52,7 @@ export const ButtonLabel = styled.Text`
   text-align: center;
   letter-spacing: 0.01px;
   text-transform: uppercase;
-  color: ${Colors.lightest};
+  color: ${props => (props.color ? props.color : Colors.lightest)};
```
Comment (thumn) [inline components/BaseComponents.js]: try `color: ${props => props.color || Colors.lightest};`?

File: screens/CheckoutScreen.js [hunk 4]
```diff
@@ -1,112 +1,108 @@
...
     });
   }
 
+  applyRewardsCallback = (rewardsApplied, totalPrice) => {
+    // Update rewards in parent state
+    this.setState({ rewardsApplied, totalPrice });
+  };
+
+  updateQuantityCallback = (product, quantity, priceDifference) => {
+    console.log(priceDifference);
```
Comment (thumn) [inline screens/CheckoutScreen.js]: remove? :0

Comment (anniero98) [reply screens/CheckoutScreen.js]: oops, ack

File: styled/modal.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  margin: auto;
+  justify-content: space-around;
+  align-items: center;
+  background-color: white;
+  top: -200px;
+`;
+
+export const QuantityInput = styled.TextInput.attrs({
+  keyboardType: 'numeric',
+  maxLength: 3
```
[the remark this answers is not in the document]
Comment (anniero98) [reply styled/modal.js]: Fixed in `rewards-redesign`, but copying changes over for now

File: components/BaseComponents.js (hunk 3 above)
Comment (anniero98) [reply components/BaseComponents.js]: ack!

**Comment by anniero98** [comment]
> ### UX/Flow Issues
> * Upon opening the QTY modal, set the focus to the text field so the keyboard pops up without having to click it.
>   
>   * If the particular product hasn't been added before, default the field to be blank (so the clerk can just directly begin typing, not have to backspace the 0). If they are modifying the quantity, keep the field with the existing value.
>   * With the keyboard up, this would also reduce the awkwardness of having the modal positioned high up on the screen (and we hopefully won't have to worry about KeyboardAvoidingView at all.
> * The same issue I noted on Tommy's login PR — having to click twice (once to dismiss the keyboard, again to click the button) could be quite annoying (this also connects to the note above! it helps if you can directly click the button without having to dismiss the keyboard)

## Partially addressed
Fixed the start value (didn't consider that while setting defaults). Not going to address the rest in this PR. Please note for future tasks!

> ### Bugs/Edge Cases
> **Bug: Updating quantity with an empty text field (this is probably the only urgent/likely-to-actually-occur bug that should be fixed in this PR)**
> 
> * **To recreate:** Add 10 blueberries, then click on the line item, clear the text field and hit update quantity.
> * **Result**: Order total goes to $NaN, and clicking on the Blueberries product card shows the quantity till as 10. If I update Blueberries qty to 0 and hit update, Order Total stays NaN

## Fixed (possibly temp.)
Fixed this by disabling the button if `currentQuantity` is an empty string.

Initially I just added a check and allowed the modal to be closed; however, I don't think it'd be expected behavior to be able to enter an empty string and update quantity, so if we **do** enable the button even for empty strings...
I think we have two options: add an error message ("Please enter a quantity") or we should have that act as if a quantity of 0 was entered. I think the latter is a little more intuitive but doesn't make sense logically; anyway, right now, I'm just disabling it; they don't get stuck because they just need to enter some number, or exit the modal. 

So I think we can decide later.

Note that this would have had issues (exiting while editing a LineItemComponent) if I hadn't also fixed the "exit without saving" props re-rendering bug.

> **Bug: Negative Quantities**
> 
> * **To recreate:** Add 10 blueberries the proper way, then add a NEGATIVE value of another item:
> * **Result**: the negative item does not get added as a line item, but the cart total still decreases.

> **Bug: NaN quantities**
> 
> * **To recreate:** Click a product, enter a NaN qty (punctuation or something)
> * **Result:** The NaN input error is caught and logged in the console, but there's no visual indication or anything that blocks the input, and the Order Total goes to NaN (and there's no way to undo, so when I check out with a NaN cart and point total, I get a big red error 'undefined is not an object'

## Fixed (possibly temp.)
Fixed both of these by only allowing input to be integers;  it just does nothing unless an integer is entered. I think this is easier than parsing bad input, but this is probably something that Ace should consider (i.e if this is the best behavior to enforce).

Also I think the negative thing is really amusing but a very good catch! Bug is from a combo of how cart is updated + when line items are shown; pretty subtle.

> **Bug: Closing the modal**
> 
> * **To recreate:** click Broccoli Crowns card, enter qty 8 but don't hit UPDATE, instead hit X.
> * **Result:** it doesn't add as a line item, but when I click Broccoli Crowns again, the qty is pre-filled with 8.
> * **Expected behavior:** probably should clear the field or reset to placeholder text on hitting X (I could see this getting confusing if they are trying to add again but think the filled text indicates that they already added it and then look for it in line items but it's not there so get more confused..)

## Fixed (possibly temp.)
I considered this and tried a couple of fixes, but after running through some scenarios I think it should actually always show the current quantity upon open. I added some conditionals in `setModalVisible` to handle this.

> **Bug: Rewards Button remains active with insufficient points (might be fixed in your rewards PR)**
Pretty sure this is fixed in `rewards-redesign` - I changed the condition. I noted it and will check before I make that PR.

A lot of these are design-dependent, so possibly will change in the future, but I think UI is at least logically correct now.

File: screens/CheckoutScreen.js [hunk 6]
```diff
@@ -1,113 +1,108 @@
...
     // in handleSubmit() for updated amount.
     msg = msg.concat(`\nTotal Price: $${this.state.totalPrice.toFixed(2)}\n`);
     msg = msg.concat(`Total Points Earned: ${totalPoints}`);
     return msg;
-  }
+  };
 
   // Adds the transaction to the user's account and updates their points.
-  async confirmTransaction() {
+  confirmTransaction = async () => {
```
Comment (aivantg) [inline screens/CheckoutScreen.js]: Hehe i <3 arrow functions, let’s get it on a shirt

File: screens/CheckoutScreen.js [hunk 7]
```diff
@@ -141,111 +136,58 @@ export default class CheckoutScreen extends React.Component {
     Alert.alert('Confirm Transaction', this.generateConfirmationMessage(totalPoints), [
```
Comment (aivantg) [inline screens/CheckoutScreen.js]: This is a pretty cool Alert system :o

File: screens/CheckoutScreen.js [hunk 8]
```diff
@@ -141,111 +136,58 @@ export default class CheckoutScreen extends React.Component {
...
             <Subhead>Current Sale</Subhead>
             {/* Cart container */}
             <View style={{ height: '40%', paddingBottom: '5%' }}>
               <ScrollView>
-                {cart.map(product => (
-                  <TouchableOpacity key={product.id} onPress={() => this.removeFromCart(product)}>
-                    <ProductCartCard product={product} />
-                  </TouchableOpacity>
-                ))}
+                {Object.entries(cart).map(([id, product]) => {
```
Comment (aivantg) [inline screens/CheckoutScreen.js]: Damn `Object.entries` + array destructuring, insane 👏

File: screens/CheckoutScreen.js [hunk 9]
```diff
@@ -141,111 +136,58 @@ export default class CheckoutScreen extends React.Component {
...
-              keyExtractor={product => product.id}
-              numColumns={5}
-              data={products}
-              renderItem={({ item }) => (
-                <TouchableOpacity onPress={() => this.addToCart(item)}>
-                  <ProductDisplayCard product={item} />
-                </TouchableOpacity>
-              )}
-            />
+            {Object.entries(cart).map(([id, product]) => (
```
Comment (aivantg) [inline screens/CheckoutScreen.js]: This is not a genuine nit since it’s so small, just a thought out loud. Since this mapping is used twice, a decent opportunity to create a functional component that takes in `cart` and `isLineItem`. 

Again, not actually recommending cuz it’s an annoying inconsequential change, but it makes me wonder if there is a future in React with a ton of mini functional components.....i don’t like functional components 😅😂

A middle ground I at some point learned (that Ethan hates) is the idea of creating a function along the lines of: 
```
renderProductList(cart, isLineItem) { 
  return Object.entries(cart)...
}
```
 
And then calling `this.renderProductList(cart, true)`. It’s like a mini functional component that lives in the same file? Yeah idk what my thoughts are on it

File: screens/modals/QuantityModal.js [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+      modalVisible: false,
+      product: null,
+      // Needs to be string type because it's TextInput
+      currentQuantity: '',
+      isLoading: true
+    };
+  }
+
+  componentDidMount() {
+    const { product } = this.props;
```
Comment (aivantg) [inline screens/modals/QuantityModal.js]: Don’t you generally _not_ use prop destructuring in your codebase? Why bother?

File: screens/modals/QuantityModal.js [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+    this.resetCurrentQuantity();
+    this.setState({
+      product,
+      isLoading: false
+    });
+  }
+
+  // Forces a re-render when new props are passed
+  componentWillReceiveProps(nextProps) {
```
Comment (aivantg) [inline screens/modals/QuantityModal.js]: Interesting, did not know this function was a thing! What’s the difference between this and `componentDidUpdate` do you know?

File: screens/modals/QuantityModal.js [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+    const newQuantity = nextProps.product.quantity;
+    if (this.state.currentQuantity !== newQuantity) {
+      this.setState(prevState => ({
```
Comment (aivantg) [inline screens/modals/QuantityModal.js]: It seems like you’re doing a very similar thing in `componentDidMount` and `componentWillReceiveProps` right now: taking props and setting up state with it. You should keep that as one function and then call that function in both of the lifecycle methods.

Pretty common practice, especially when you need to make API call to update based on new props and all, 1 function for the call, call it in both lifecycle methods

File: screens/modals/QuantityModal.js (hunk 11 above)
Comment (aivantg) [reply screens/modals/QuantityModal.js]: Apparently the `willReceiveProps` is deprecated LOL

File: screens/modals/QuantityModal.js [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+  resetCurrentQuantity = () => {
+    const { product } = this.props;
+    const quantityInt = product.quantity;
+    this.setState({ currentQuantity: quantityInt === 0 ? '' : quantityInt.toString() });
+  };
+
+  setModalVisible = visible => {
+    if (visible) {
+      this.resetCurrentQuantity();
```
Comment (aivantg) [inline screens/modals/QuantityModal.js]: Oh shit, did not see that `resetCurrentQuantity` is used here too. I think you could still organize it into one function that accomplishes all goals to repeat less, but honestly i think this is fine too

Comment (aivantg) [reply screens/modals/QuantityModal.js]: Something like a function called`resetState(product)`

File: screens/modals/QuantityModal.js [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+    this.setModalVisible(!this.state.modalVisible);
+  };
+
+  // Communicate to parent component
+  handleUpdateCart = () => {
+    const initialQuantity = this.state.product.quantity;
+    const currentQuantityInt = parseInt(this.state.currentQuantity, 10);
+    const priceDifference = (currentQuantityInt - initialQuantity) * this.state.product.customerCost;
+    this.props.callback(this.state.product, currentQuantityInt, priceDifference);
+    this.setModalVisible(!this.state.modalVisible);
```
Comment (aivantg) [inline screens/modals/QuantityModal.js]: Is there ever a case where this function would _not_ want to set `modalVisible` to `false`? Toggling it _should_ have that affect always, but just saying `this.setModalVisible(false)` might be more explicit

File: screens/modals/QuantityModal.js [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+
+  setModalVisible = visible => {
+    if (visible) {
+      this.resetCurrentQuantity();
+    }
+    this.setState({ modalVisible: visible });
+  };
+
+  handleShowModal = () => {
+    this.setModalVisible(!this.state.modalVisible);
```
Comment (aivantg) [inline screens/modals/QuantityModal.js]: Again, any reason not to do `this.setModalVisible(true)` if this is always showing the modal (and not toggling it). I have NO idea how the UI works so please correct me!

Comment (aivantg) [reply screens/modals/QuantityModal.js]: I might just suggest renaming to `toggleModal` as a more literal name

File: screens/modals/RewardModal.js [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+    this.setState({
+      rewardsApplied,
+      rewardStatus,
+      totalPrice,
+      isLoading: false
+    });
+  }
+
+  componentWillReceiveProps({ totalPrice }) {
+    this.setState(prevState => ({ ...prevState, totalPrice }));
```
Comment (aivantg) [inline screens/modals/RewardModal.js]: Veeeery slick

File: screens/modals/RewardModal.js [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+      rewardsApplied: 0,
+      rewardStatus: [],
+      totalPrice: 0,
+      isLoading: true
+    };
+  }
+
+  componentDidMount() {
+    const { rewardsAvailable, rewardsApplied, totalPrice } = this.props;
+    const rewardStatus = [];
```
Comment (aivantg) [inline screens/modals/RewardModal.js]: Dumb that you can’t do something like : `rewardStatus = range(0, rewardsAvailable).map(...)`

You can with external libraries or unreadable one-liners but it should be in vanilla JavaScript 😠

File: screens/modals/RewardModal.js [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+    if ((totalPrice < 5 && rewardsApplied === 0) || rewardsAvailable === 0) {
+      return;
+    }
+    this.setModalVisible(!this.state.modalVisible);
+  };
+
+  handleApplyRewards = () => {
+    // Communicate to parent component
+    this.props.callback(this.state.rewardsApplied, this.state.totalPrice);
+    this.setModalVisible(!this.state.modalVisible);
```
Comment (aivantg) [inline screens/modals/RewardModal.js]: Same comment as earlier modal, is toggle actually necessary or would `this.setModalVisible(false)` suffice

File: screens/modals/RewardModal.js (hunk 17 above)
Comment (aivantg) [reply screens/modals/RewardModal.js]: Okay so i chased this rabbit hole. One other way you couuuld do this is: 

```
const rewardStatus = new Array(rewardsAvailable).fill(false).map((_, i) => (...));
```

Buuuut maybe we’re getting less legible and too technical here 😂

File: screens/modals/QuantityModal.js (hunk 11 above)
Comment (anniero98) [reply screens/modals/QuantityModal.js]: "I won't look at the comments until later" chronically lying about not .. coding. Damn I didn't realize 😂 https://blog.bitsrc.io/react-16-lifecycle-methods-how-and-when-to-use-them-f4ad31fb2282 I looked at the old version of this blog post LOL -- anyway, I think it's been replaced by `getDerivedStateFromProps` which is.. an incredibly specific name and exactly what I'm doing.

Also, the fact that the author does not seem to like this function 😂
> Even with the above cases, there’s usually a better way to do it. But getDerivedStateFromProps will have your back when worst comes to worst.

 I think `componentDidUpdate` doesn't seem to be the same use case? So I'm going to go ahead and replace it with `getDerivedStateFromProps`  ..

I shall change this at minimum in the rewards redesign PR, which apparently will become an updating modals PR (it's tentatively done; maybe I'll make a draft PR after finishing HW https://github.com/calblueprint/dccentralkitchen-clerks/compare/annie/quantity-modal...annie/reward-redesign)

File: screens/modals/QuantityModal.js (hunk 15 above)
Comment (anniero98) [reply screens/modals/QuantityModal.js]: This is a good point, and I was wondering myself... I actually already renamed `handleShowModal` to `toggleVisibility` in the next PR (LOL), will consider just replacing `setModalVisible` altogether. Prettyyyy sure the `handleShowModal()` was a) from boilerplate and b) used with conditions when I didn't know `disabled` was a prop for buttons (^:

File: screens/modals/RewardModal.js (hunk 17 above)
Comment (anniero98) [reply screens/modals/RewardModal.js]: Okay wait so I actually tried this before (we've been iterating over a range set by some int for awhile now); forget that this status array is deprecated with the redesign for a second... 🤣 . BUT for some reason I could NOT understand, I would always get this "Array size is not a small enough positive integer" error _non-deterministically_ (dear lord) no matter how I initialized the integer, so I gave up and used a for loop ;__;

File: screens/CheckoutScreen.js [hunk 19]
```diff
@@ -1,113 +1,108 @@
...
   }
 
+  applyRewardsCallback = (rewardsApplied, totalPrice) => {
+    // Update rewards in parent state
+    this.setState({ rewardsApplied, totalPrice });
+  };
+
+  updateQuantityCallback = (product, quantity, priceDifference) => {
+    this.setState(prevState => ({
+      cart: update(prevState.cart, { [product.id]: { quantity: { $set: quantity } } }),
```
[the remark this answers is not in the document]
Comment (anniero98) [reply screens/CheckoutScreen.js]: I asked about this in Slack, and to this day I still am not quite sure why the vanilla js way of writing it didn't in fact update the state (I tried a couple ways of writing the update) - luckily `update` from `react-addons-update` seems to be recommended by the docs, and it worked, so I just went with it :/

File: screens/CheckoutScreen.js (hunk 9 above)
Comment (anniero98) [reply screens/CheckoutScreen.js]: HAHAHA. I'm totally going to do this because I think it's kind of unreadable as it is right now 🤠
